DNET-532

zabulus opened the pull request:
1. I've added test for culture aware comparison of parameter names.
2. I've added fallback to ordinal comparison in case parameter names don't have unicode characters.
3. Also added a little improvement to Remove method. It removes double comparison via Contains before deleting element from list.

About second I have doubts, the implementation is not very clean. Though, for original case difference between "by index" vs "by name" is about 4 seconds.

zabulus commented on NETProvider/src/FirebirdSql.Data.UnitTests/FbParameterCollectionTests.cs:
```diff
@@ -23,19 +23,9 @@
-	[TestFixture(FbServerType.Default)]
-	[TestFixture(FbServerType.Embedded)]
-	public class FbParameterCollectionTests : TestsBase
+	[TestFixture]
```
For this tests connection is not necessary

zabulus commented on NETProvider/src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbParameterCollection.cs:
```diff
@@ -90,6 +91,21 @@ public override object SyncRoot
 			get { return ((ICollection)_parameters).SyncRoot; }
 		}
 
+		internal bool? _collectionHasParameterWithUnicodeName;
```
these fields are internal for tests

cincuranet replied:
Isn't it enough to have property available?

zabulus replied:
in tests I've also checked that flag is nullified correctly when parameters changed, the property doesn't allow such check. They both, field and property, are implementation details.

cincuranet replied:
I feel that checking that field is too much. For instance if we recompute the value with every change it will still be correct but the test will assume it's not.

zabulus replied:
I don't want to recompute field on each change. I intentionally made computing the field only on-demand if ```IndexOf``` is called.

zabulus replied:
I think I can make the property private, but the field still need to be internal for tests

cincuranet replied:
> I don't want to recompute field on each change. I intentionally made computing the field only on-demand if IndexOf is called.

That was just an example. You may as well use two properties, one with value and one saying whether the value needs recomputing or not. What I'm saying is that the internal representation should not matter. Only the desired outcome of behavior.

I don't mind if you don't like it. I'll merge it anyway and then I'll change it. So it may as well stay as is.

> I think I can make the property private, but the field still need to be internal for tests

That's even worse.

zabulus replied:
> That was just an example. You may as well use two properties, one with value and one saying whether the value needs recomputing or not. What I'm saying is that the internal representation should not matter. Only the desired outcome of behavior.

Yes that may work.

cincuranet commented on NETProvider/src/FirebirdSql.Data.UnitTests/FbParameterCollectionTests.cs:
```diff
@@ -49,6 +39,64 @@ public void AddTest()
+		}
+
+		[Test]
+		[Category("Run")]
```
`Run` category?

zabulus replied:
Need to remove

cincuranet commented on NETProvider/src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbParameterCollection.cs:
```diff
@@ -239,13 +266,13 @@ public override void Insert(int index, object value)
 		public void Remove(FbParameter value)
 		{
-			if (!Contains(value))
+			if(!_parameters.Remove(value))
```
:+1:

cincuranet commented on NETProvider/src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbParameterCollection.cs:
```diff
@@ -214,20 +231,30 @@ public override int IndexOf(string parameterName)
+			bool parameterUnicodeName = Encoding.UTF8.GetByteCount(parameterName) != parameterName.Length;
+			bool collectionHasParameterWithUnicodeName = CollectionHasParameterWithUnicodeName;
+			StringComparison usedComparison = StringComparison.OrdinalIgnoreCase;
+			if(parameterUnicodeName || collectionHasParameterWithUnicodeName)
```
Minor formatting mistake. :)

zabulus replied:
could you clarify?

cincuranet replied:
Space after `if`.

cincuranet commented:
Found a case where it fails:
```
var collection = new FbParameterCollection();
var parameter = new FbParameter() { ParameterName = "test" };
collection.Add(parameter);
Assert.IsFalse(collection.CollectionHasParameterWithUnicodeName);
parameter.ParameterName = "řčšřčšřčš";
Assert.IsTrue(parameter.IsUnicodeParameterName);
Assert.IsTrue(collection.CollectionHasParameterWithUnicodeName);
```

zabulus commented:
Your case lacks ```IndexOf``` call. Only then field and property are reevaluated.
I have doubts about all this optmization. The case you wrote is a good evidence that all this stuff with field/property is pretty obscure.

cincuranet commented:
> Your case lacks IndexOf call. Only then field and property are reevaluated.

It fails anyway. ;)
```
var collection = new FbParameterCollection();
var parameter = new FbParameter() { ParameterName = "test" };
collection.Add(parameter);
var dummy1 = collection.IndexOf("dummy");
Assert.IsFalse(collection.CollectionHasParameterWithUnicodeName);
parameter.ParameterName = "řčšřčšřčš";
var dummy2 = collection.IndexOf("dummy");
Assert.IsTrue(parameter.IsUnicodeParameterName);
Assert.IsTrue(collection.CollectionHasParameterWithUnicodeName);
```

> I have doubts about all this optmization. The case you wrote is a good evidence that all this stuff with field/property is pretty obscure.

I agree. But you started the PR, so I though I'll give it a try.

zabulus commented:
> It fails anyway. ;)
I got why it fails. I didn't consider the case, where parameter name can be changed outside collection. I think I'll revert adding the field instead use only property, so it at least will be correct.

zabulus commented:
Removed field. Property just evaluates ```.Any()```.
I think I can implement optimization for this in other way. Something like ```version``` in ```List``` implementation. It should incremented when underlying list is changed or any element's ```ParameterName``` property chaged.

cincuranet commented:
> Removed field. Property just evaluates .Any().

It would be interesting to see (aka measure) what we're gaining now. And how behaves with 1, 10, 100, ... parameters in collection. Maybe the `Any` + ordinal compare would not bring desired benefit over culture aware compare.

> I think I can implement optimization for this in other way. Something like version in List implementation. It should incremented when underlying list is changed or any element's ParameterName property chaged.

That's again on the edge.

zabulus commented:
For case described in the issue difference is about 7-8 secs. Looks better than culture aware comparison.

cincuranet commented:
I see a different values. Most visible it's in worst case scenario. When the collection contains only ASCII parameter names, then the `Any` takes a lot of time offsetting the ordinal compare. I see about two orders of magnitude difference on a 100 items collection with 100 compares.

zabulus commented:
Ok I've added new implementation version. 10.59 (IBN) vs 6.11 (IBI)
This introduces ```version``` field that increments when changes of internal list or if field ```ParameterName``` of any FbParameter instance change occurs.
Also I've added test around Parent field. There were several issues about it.
It was never set by a ```Collection``` and it wasn't checked in insert method.

BTW, I think that it's getter can freely be public, though it's not part of ```DbParameter``` contract.

cincuranet commented:
OK, let me first pick the `Parent` fixes. These are going in, for sure. I'll commit these in a minute.

Why the `version`? Isn't simple `bool` flag enough? You just need to know whether the value computed is stale or not.

zabulus commented:
@cincuranet something left for this?

cincuranet commented:
Yes. This:
> Why the version? Isn't simple bool flag enough? You just need to know whether the value computed is stale or not.

zabulus commented:
I've fixed it in last commit

cincuranet commented:
Ahh, sorry. I missed that.
